### PR TITLE
BUG: fix importing numpy in Python's optimized mode

### LINKED ISSUE #27866: BUG: NumPy 2.2.0rc1 import fails with PYTHONOPTIMIZE=2
A build failure was observed per https://github.com/scipy/scipy/issues/21957 for this particular SciPy build incantation: `python -m build --wheel --no-isolation -Ccompile-args=-j32`

I bisected to the NumPy commit below. Is this a bug in NumPy or something we need to adjust downstream? @mhvk and @seberg were the PR reviewers.

```
b47bd25e3caba52321a6ba420650fc0eff97f473 is the first bad commit
commit b47bd25e3caba52321a6ba420650fc0eff97f473
Author: Clément Robert <cr52@protonmail.com>
Date:   Fri Nov 22 11:56:31 2024 +0100

    API: register NEP 35 functions as array_functions (#27490)
    
    Register `like=` functions as array-functions.   This is currently decided based on finding the magic `${ARRAY_FUNCTION_LIKE}` string in the documentation.
    
    Close gh-27451
```

### PR DESCRIPTION
**Pull request by neutrinoceros**
Follow up to gh-27490, fix gh-27866

### CODE REVIEW

File: numpy/_core/overrides.py [hunk 1]
```diff
@@ -21,6 +22,11 @@
 
 def get_array_function_like_doc(public_api, docstring_template=None):
```
Comment (neutrinoceros) [inline numpy/_core/overrides.py:23]: I realized that changing the default value for docstring_template to an empty string would probably work too, but I cannot test it right now (afk).

**Review by mhvk — COMMENTED** [review]
This looks good but perhaps good to follow https://github.com/numpy/numpy/issues/27866#issuecomment-2503031137 and also change a CI test to ensure this does not recur.

**Comment by neutrinoceros** [comment]
sure thing. What for would be appropriate ? Do we want to run the whole test suite or just try `python -OO -c 'import numpy'` ? and where should this new job live ?

**Comment by rgommers** [comment]
The whole test suite, and no new job needed I think - just adding `PYTHONOPTIMIZE=2` to the existing `full` job at https://github.com/numpy/numpy/blob/main/.github/workflows/linux.yml#L153 should cover our bases.

**Comment by neutrinoceros** [comment]
Interesting, this found one failing test under PYTHONOPTIMIZE=2 . Let me try to fix this too.

**Comment by neutrinoceros** [comment]
Only the test itself didn't support optimized mode because it relied on having IO operations evaled through assert statements. Separating evaluation from assertion should do the trick (it works locally, at least).

**Comment by mhvk** [comment]
Does this mean that all the `assert` in the tests are actually removed? I.e., that `full` now only tests that the tests run, but not actually that the output is correct? That may not be horrible, but perhaps not quite intended either... Especially as surely there are other tests whether something is calculated inside the `assert`, and thus coverage may be incomplete too.

I wonder if we can pick a shorter run and just do tests twice, as in
```
pytest ...
PYTHONOPTIMIZE=2 pytest ...
```
(Checking this locally on another project, I get a failure because astropy can also not be loaded with `-O`...)

Alternatively, perhaps we do need a special branch that just imports everything?

**Comment by neutrinoceros** [comment]
> Does this mean that all the assert in the tests are actually removed? 

yes. That is a direct expectation from running python with `PYTHONOPTIMIZE` >0

**Comment by rgommers** [comment]
Wasn't that fixed? IIRC that was a main reason for us avoiding plain `assert` in the past (in favor of our own `np.testing.assert_`), and `pytest` got robust against that at some point I believe.

**Comment by neutrinoceros** [comment]
Oh yes, sorry, I see from doing more experiments with it locally that actually, only `assert` statements that are not part of tests are expected to be skipped. In fact, pytest warns about it
```
PytestConfigWarning: assertions not in test modules or plugins will be ignored because assert statements are not executed by the underlying Python interpreter (are you using python -O?)
```
though, I don't understand why that one test needed tweaking...

**Review by mhvk — APPROVED** [review]
It was the test tweak that made me wonder - I also thought this was fixed. Just to be sure, I also confirmed that a failing test actually fails with optimization on.

Anyway, chalk it up to the mysteries of python...

**Comment by mhvk** [comment]
For reference, `pytest` indeed rewrites the `assert` upon import of the test module - https://docs.pytest.org/en/stable/announce/release-2.1.0.html and https://docs.pytest.org/en/stable/how-to/assert.html#assert-details. I'm *still* not sure why the test needed to be rewritten, though!

**Comment by seberg** [comment]
My guess is that this is due to:
```
numpy/_core/tests/test_deprecations.py:        from numpy.lib.tests.test_io import TextIO
```
i.e. the module is already imported before pytest does its magic.  I think it would be nice if pytest would warn (or even error) but dunno if that is doable.

Maybe open an issue there?

So an alternative fix would be moving the `TextIO` use (or seeing if the deprecation test is even worth it).

But, since that seems the reason, this work-around is also not a big deal.

**Review by rgommers — APPROVED** [review]
This seems to do the job, so while we're pondering the `pytest` magic related hiccups, I'll go ahead and merge this and rebuild the nightly wheels to avoid more issues downstream (the [failing SciPy CI job](https://github.com/scipy/scipy/pull/21961) needs a local fix though, since it currently pulls in numpy 2.2.0rc1).

**Comment by rgommers** [comment]
Thanks @neutrinoceros, @mhvk and @seberg!

**Pull request merged**